c/tskit/trees.c

```diff
@@ -8145,3 +8165,339 @@ tsk_treeseq_extend_edges(const tsk_treeseq_t *self, int max_iter,
     tsk_table_collection_free(&tables);
     return ret;
 }
+
+/* ======================================================== *
+ * Pair coalescence
+ * ======================================================== */
+
+typedef struct {
+    /* TODO: better to include the implm from haplotype_matching.c, rather than
+     * duplicating? */
+    double value;
+    tsk_size_t index;
+} tsk_node_argsort_t;
+
+static int
+cmp_node_argsort(const void *a, const void *b)
+{
+    const tsk_node_argsort_t *ia = (const tsk_node_argsort_t *) a;
+    const tsk_node_argsort_t *ib = (const tsk_node_argsort_t *) b;
+    int ret = (ia->value > ib->value) - (ia->value < ib->value);
+    if (ret == 0) {
+        ret = (ia->index > ib->index) - (ia->index < ib->index);
+    }
+    return ret;
+}
+
+static int
+get_time_windows_index_map(const tsk_treeseq_t *self, tsk_size_t *num_time_windows,
+    const double *time_windows, tsk_id_t *result)
+{
+    int ret = 0;
+    const tsk_table_collection_t *tables = self->tables;
+    const tsk_size_t num_nodes = tables->nodes.num_rows;
+    tsk_node_argsort_t *nodes_time = NULL;
+    tsk_size_t i, j;
+
+    if (*num_time_windows == 0) { /* keep nodes unpooled */
+        for (i = 0; i != num_nodes; i++) {
+            result[i] = (tsk_id_t) i;
+        }
+        *num_time_windows = num_nodes;
+        goto out;
+    }
+
+    nodes_time = tsk_malloc(num_nodes * sizeof(*nodes_time));
+    if (nodes_time == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+    for (i = 0; i != num_nodes; i++) {
+        result[i] = TSK_NULL; /* nodes outside windows */
+        nodes_time[i].value = tables->nodes.time[i];
+        nodes_time[i].index = i;
+    }
+    qsort(nodes_time, (size_t) num_nodes, sizeof(*nodes_time), cmp_node_argsort);
+
+    for (i = 0, j = 0; i != *num_time_windows; i++) {
+        while (j != num_nodes && nodes_time[j].value < time_windows[i + 1]) {
+            if (nodes_time[j].value >= time_windows[i]) {
+                result[nodes_time[j].index] = (tsk_id_t) i; /* nodes inside windows */
+            }
+            j += 1;
+        }
+    }
+out:
+    tsk_safe_free(nodes_time);
+    return ret;
+}
+
+int
+tsk_treeseq_pair_coalescence_stat(const tsk_treeseq_t *self, tsk_size_t num_sample_sets,
+    const tsk_size_t *sample_set_sizes, const tsk_id_t *sample_sets,
+    tsk_size_t num_set_indexes, const tsk_id_t *set_indexes, tsk_size_t num_windows,
+    const double *windows, tsk_size_t num_time_windows, const double *time_windows,
+    tsk_flags_t options, double *result)
+{
+    int ret = 0;
+    double left, right, remaining_span, window_span;
+    tsk_id_t e, p, c, u, v, n, w, i, j, k, row, col, inp;
+    tsk_size_t total_samples;
+    int weight;
```

Comment:
Plain 'int' values are tricky, as they can be anything from 16 bit to 64 bit depending on platform. It saves headaches to be explicit about the types, if there's any chance at all of overflow. So here - what's the maximum weight value? I'd err on the side of caution and declare things as int64, if there's any possibility of overflow from int32.

Reply:
The reason I'd settled on int rather than tsk_id_t (int64) for these, is that there's a compiler warning/error due to precision loss when multiplying int64 weight with double below. I think if I explicitly cast from int64 to double I can avoid the warnings-- or are there issues with this?

Reply:
Cast int64 to double then, that's definitely safe in the ranges we'll be dealing with